src/main/java/reactor/ipc/netty/channel/ChannelOperationsHandler.java

```diff
@@ -360,50 +378,70 @@ void drain() {
 
 				v = pendingWrites.poll();
 
-				if (v instanceof Publisher) {
-					Publisher<?> p = (Publisher<?>) v;
-
-					if (p instanceof Callable) {
-						@SuppressWarnings("unchecked") Callable<?> supplier =
-								(Callable<?>) p;
-
-						Object vr;
-
-						try {
-							vr = supplier.call();
-						}
-						catch (Throwable e) {
-							promise.setFailure(e);
-							continue;
-						}
-
-						if (vr == null) {
-							promise.setSuccess();
-							continue;
+				if (!innerActive && v instanceof PendingWritesOnCompletion) {
+					boolean last = pendingWrites.isEmpty();
+					if (!future.isDone() && hasPendingWriteBytes()) {
+						ctx.flush();
+						if (!future.isDone() && hasPendingWriteBytes()) {
+							pendingWriteOffer.test(future, v);
 						}
-
-						if (inner.unbounded) {
-							doWrite(vr, promise, null);
+					}
+					if (last && WIP.decrementAndGet(this) == 0) {
+						break;
+					}
+				}
+				else if (future instanceof ChannelPromise) {
+					ChannelPromise promise = (ChannelPromise) future;
+					if (v instanceof Publisher) {
+						Publisher<?> p = (Publisher<?>) v;
+	
+						if (p instanceof Callable) {
+							@SuppressWarnings("unchecked") Callable<?> supplier =
+									(Callable<?>) p;
+	
+							Object vr;
+	
+							try {
+								vr = supplier.call();
+							}
+							catch (Throwable e) {
+								promise.setFailure(e);
+								continue;
+							}
+	
+							if (vr == null) {
+								promise.setSuccess();
+								continue;
+							}
+	
+							if (inner.unbounded) {
+								doWrite(vr, promise, null);
+							}
+							else {
+								innerActive = true;
+								inner.promise = promise;
+								inner.onSubscribe(Operators.scalarSubscription(inner, vr));
+							}
 						}
 						else {
 							innerActive = true;
 							inner.promise = promise;
-							inner.onSubscribe(Operators.scalarSubscription(inner, vr));
+							p.subscribe(inner);
 						}
 					}
 					else {
-						innerActive = true;
-						inner.promise = promise;
-						p.subscribe(inner);
+						doWrite(v, promise, null);
 					}
 				}
-				else {
-					doWrite(v, promise, null);
-				}
 			}
 		}
 	}
 
+	private boolean hasPendingWriteBytes() {
+		ChannelOutboundBuffer outBuffer = ctx.channel().unsafe().outboundBuffer();
```

Comment:
may keep the reference when channel is seen first time in a class property vs evaluating here.